### PR TITLE
Slugify and Canonicals

### LINKED ISSUE #285: Implement slugs and canonicals

### PR DESCRIPTION
**Pull request by Jordan141**
This fixes #285, #283, #282

Goals of this PR:
- Install Slugify
- Create article and author links based on slugified article title/ user fullname respectively
- Store outdated slugified links and redirect to current link
- Check for duplicates, if the slug is a duplicate link, add an incremented number at the end and check it.
- Change Joi schemas to reflect new content rule changes
- Research canonicals and apply them.

### CODE REVIEW

File: app/app.js [hunk 1]
```diff
@@ -153,6 +154,7 @@ app.use(async (req, res, next) => {
         res.locals.searchTerm = ""
         res.locals.commonCategories = await utils.findCommonCategories()
         res.locals.websiteUrl = req.hostname
+        res.locals.cannonicalUrl = req.protocol + '://' + req.get('host') + req.originalUrl
```
Comment (bilgorajskim) [inline app/app.js:157]: typo

File: app/models/article.js [hunk 2]
```diff
@@ -40,5 +44,22 @@ function categoryValidation(val) {
...
+        Link.create({link: newLink, docType, doc_id: docId})
+        return link
+    }
+    return await getLink(link, docType, docId, n + 1)
+}
+
+articleSchema.pre('validate', async function(next) {
+    if(!this.isModified('title')) return next()
+    const sluggedLink = slugify(this.title, SLUGIFY_OPTIONS)
+    this.link = await getLink(sluggedLink, ARTICLE_TYPE, this._id)
```
Comment (michaelKurowski) [inline app/models/article.js:60]: can `getnLink` crash? should we try...catch it?

File: app/models/user.js [hunk 3]
```diff
@@ -40,4 +43,21 @@ userSchema.plugin(passportLocalMongoose, {
     }
 })
 
+async function getLink(link, docType, docId, n = 1) {
```
Comment (michaelKurowski) [inline app/models/user.js:46]: Looks like you have duplicated `getLink` function

File: app/models/user.js [hunk 4]
```diff
@@ -40,4 +43,21 @@ userSchema.plugin(passportLocalMongoose, {
...
+        Link.create({link: newLink, docType, doc_id: docId})
+        return link
+    }
+    return await getLink(link, docType, docId, n + 1)
+}
+
+userSchema.pre('validate', async function(next) {
+    if(!this.isModified('fullname')) return next()
+    const sluggedLink = slugify(this.fullname, SLUGIFY_OPTIONS)
+    this.link = await getLink(sluggedLink, USER_TYPE, this._id)
```
Comment (michaelKurowski) [inline app/models/user.js:59]: Shouldn't we try...catch it?

Comment (michaelKurowski) [reply app/models/user.js:59]: Should we create a link during validation? Validation sounds like a thing that shouldn't modify the data inputted inside.

File: app/models/article.js (hunk 2 above)
Comment (michaelKurowski) [reply app/models/article.js:60]: Should we create a link during validation? Validation sounds like a thing that shouldn't modify the data inputted inside.

File: app/routes/articles.js [hunk 5]
```diff
@@ -221,9 +213,10 @@ router.put('/:link', checkArticleOwnership, validation(updateArticle, BODY), asy
         req.log('Article UPDATE Route:', req.body)
         return res.redirect('/')
     }
+
     try {
-        const article = await Article.findOneAndUpdate({link: req.params.link}, {$set: req.body}, {runValidators: true}).populate('author').exec()
-        if(article.title !== req.body.title) article.link = encodeURIComponent(req.body.title.replace(SPACES, DASH))
+        let article = await Article.findOneAndUpdate({link: req.params.link}, {$set: req.body}).exec()
```
Comment (michaelKurowski) [inline app/routes/articles.js:218]: Shouldn't we try...catch this?

**Review by michaelKurowski — COMMENTED** [review]
```
web_1    | {"message":"Server is listening on 0.0.0.0:8000","level":"info"}
web_1    | {"message":"Dropping Collections...","level":"info"}
web_1    | (node:50) Warning: Accessing non-existent property 'findOne' of module exports inside circular dependency
web_1    | (Use `node --trace-warnings ...` to show where the warning was created)
web_1    | {"message":"SeedArticles Error: TypeError: Link.findOne is not a function","level":"info"}
```

**Comment by michaelKurowski** [comment]
> ```
> web_1    | {"message":"Server is listening on 0.0.0.0:8000","level":"info"}
> web_1    | {"message":"Dropping Collections...","level":"info"}
> web_1    | (node:50) Warning: Accessing non-existent property 'findOne' of module exports inside circular dependency
> web_1    | (Use `node --trace-warnings ...` to show where the warning was created)
> web_1    | {"message":"SeedArticles Error: TypeError: Link.findOne is not a function","level":"info"}
> ```

Error while reseeding app

File: app/models/user.js (hunk 3 above)
Comment (Jordan141) [reply app/models/user.js]: Fixed

File: app/models/article.js (hunk 2 above)
Comment (Jordan141) [reply app/models/article.js]: This occurs pre-validation to ensure schema property rules are enforced.

File: app/models/user.js (hunk 4 above)
Comment (Jordan141) [reply app/models/user.js]: This occurs pre-validation to ensure schema property rules are enforced.

File: app/routes/articles.js (hunk 5 above)
Comment (Jordan141) [reply app/routes/articles.js:218]: It is in one:
image

**Review by michaelKurowski — APPROVED** [review]

**Pull request merged**